File: frontend/src/lib/protocol.ts
// Mirrors backend app/protocol.py — the one JSON envelope (plan §8).

export interface Envelope<T = Record<string, unknown>> {
  type: string;
  id: string;
  payload: T;
}

// frontend -> backend
export const EDIT = "edit";
export const RUN = "run";
export const SELECT = "select";
export const CHAT = "chat";
export const ACCEPT_PATCH = "accept_patch";
export const REJECT_PATCH = "reject_patch";
export const SET_MODE = "set_mode";
export const CHECKPOINT = "checkpoint";
export const ROLLBACK = "rollback";
export const UNDO = "undo";
export const REDO = "redo";
export const PREVIEW_DIFF = "preview_diff";

// backend -> frontend
export const GEOMETRY = "geometry";
export const ERROR = "error";
export const STATUS = "status";
export const MEASUREMENT = "measurement";
export const AGENT_MESSAGE = "agent_message";
export const AGENT_PATCH = "agent_patch";
export const HISTORY = "history";
export const SOURCE = "source";

export interface Commit {
  sha: string;
  short: string;
  message: string;
  time: number;
}

export interface HistoryPayload {
  commits: Commit[];
  can_undo: boolean;
  can_redo: boolean;
}

export interface AgentMessagePayload {
  role: "assistant" | "user";
  text: string;
  error?: boolean;
}

export interface AgentPatchPayload {
  diff: string;
  rationale: string;
  targets: string[];
  new_source: string;
}

export interface Spec {
  passed: boolean;
  message: string;
}

export interface Param {
  name: string;
  value: number;
  line: number;
  is_int: boolean;
  min?: number;
  max?: number;
  step?: number;
}

// Physical-properties readout (backend app/kernel/massprops.py).
export interface PhysicalPayload {
  mass_g: number;
  volume_cm3: number;
  area_mm2: number;
  com: [number, number, number];
  inertia: number[][];
  principal: number[];
  principal_axes: number[][];
  base_z: number;
  com_height: number;
  footprint: [number, number][];
  tip_angle: number | null;
  stable: boolean;
  floats: boolean;
  effective_density: number;
  filament_g: number;
  filament_len_mm: number | null;
  cost: number;
  print_time_min: number;
  material_name: string;
  parts: number;
}

export interface GeometryPayload {
  shapes: TessShapes;
  states: Record<string, number[]>;
  bbox: BBox | null;
  ops: unknown[];
  specs: Spec[];
  params: Param[];
  stdout: string;
  stale: boolean;
  mode?: "technical" | "printability" | "highlight" | "geomdiff" | "physical";
  print_stats?: Record<string, number | string> | null;
  physical?: PhysicalPayload | null;
}

export type ViewMode = "technical" | "printability" | "highlight" | "physical";

export interface ErrorPayload {
  message: string;
  traceback: string;
  line: number | null;
}

export interface StatusPayload {
  state: "idle" | "running" | "ok" | "error";
  detail: string;
}

export type SelectKind = "face" | "edge" | "vertex" | "solid";

// Backend MEASUREMENT payload (app/kernel/select.py).
export interface MeasurementPayload {
  kind: SelectKind;
  index: number;
  selector?: string;
  selector_confidence?: "high" | "medium" | "low" | "n/a";
  description?: string;
  properties?: Record<string, unknown>;
  solid?: { bbox?: { min: number[]; max: number[]; size: number[] }; volume?: number };
  error?: string;
}

// The tessellation tree types are owned by the renderer package (single source
// of truth for geometry). Import for local use here and re-export so the WS
// protocol and the viewer agree on one definition.
import type { BBox, TessShapes, TessPart } from "@cadcode/viewer";
export type { BBox, TessShapes, TessPart };

let counter = 0;
export function makeId(): string {
  counter += 1;
  return `c${Date.now().toString(36)}-${counter}`;
}

File: backend/app/protocol.py
"""WebSocket envelope protocol (plan §8).

One JSON envelope, request/response correlated by ``id``::

    { "type": "...", "id": "uuid", "payload": { } }

Frontend -> backend:  edit · run · select · gesture · chat · accept_patch /
reject_patch · checkpoint · rollback
Backend -> frontend:  geometry · error · agent_message · agent_patch ·
measurement · status

This contract is the seam that lets the frontend and backend be built in
parallel, so it is pinned early and kept stable.
"""

from __future__ import annotations

import uuid
from typing import Any, Literal

from pydantic import BaseModel, Field

# --- message type constants -------------------------------------------------

# frontend -> backend
EDIT = "edit"
RUN = "run"
SELECT = "select"
GESTURE = "gesture"
CHAT = "chat"
ACCEPT_PATCH = "accept_patch"
REJECT_PATCH = "reject_patch"
CHECKPOINT = "checkpoint"
ROLLBACK = "rollback"
SET_MODE = "set_mode"
UNDO = "undo"
REDO = "redo"
PREVIEW_DIFF = "preview_diff"

# backend -> frontend
GEOMETRY = "geometry"
ERROR = "error"
AGENT_MESSAGE = "agent_message"
AGENT_PATCH = "agent_patch"
MEASUREMENT = "measurement"
STATUS = "status"
HISTORY = "history"
SOURCE = "source"  # backend pushes a new buffer (rollback / undo / redo)
SIMULATION = "simulation"  # motion-sim frames (per-leaf poses) + summary + specs


class Envelope(BaseModel):
    """The single wire format. ``payload`` stays loosely typed so the protocol
    can evolve without lockstep client/server model bumps; handlers validate the
    fields they need with the payload models below."""

    type: str
    id: str = Field(default_factory=lambda: uuid.uuid4().hex)
    payload: dict[str, Any] = Field(default_factory=dict)

    @classmethod
    def make(cls, type: str, payload: dict[str, Any] | None = None, id: str | None = None) -> Envelope:
        env = cls(type=type, payload=payload or {})
        if id is not None:
            env.id = id
        return env


# --- payload models (frontend -> backend) -----------------------------------


class EditPayload(BaseModel):
    """A full-source edit. The Monaco buffer *is* the new source."""

    source: str
    debounced: bool = True


class SelectPayload(BaseModel):
    """A clicked entity in the viewport (plan §7: resolve to a selector)."""

    kind: Literal["face", "edge", "vertex", "solid"]
    # path of the leaf shape that was hit (e.g. "/Group/box") + sub-index
    shape_id: str
    index: int


class ChatPayload(BaseModel):
    message: str


class CheckpointPayload(BaseModel):
    message: str = "checkpoint"


class RollbackPayload(BaseModel):
    # to a buffer-history step, an op index, or a git commit
    to: str | int


# --- payload models (backend -> frontend) ------------------------------------


class GeometryPayload(BaseModel):
    """Tessellated geometry for three-cad-viewer plus the derived op list."""

    shapes: dict[str, Any]
    states: dict[str, list[int]]
    bbox: dict[str, float] | None = None
    ops: list[dict[str, Any]] = Field(default_factory=list)
    specs: list[dict[str, Any]] = Field(default_factory=list)
    params: list[dict[str, Any]] = Field(default_factory=list)
    stdout: str = ""
    # whether this geometry is the last *good* render kept after an error
    stale: bool = False
    # "technical" (normal), "printability" (overhang heatmap), "highlight",
    # "geomdiff", or "physical" (mass/COM readout — carries `physical`, not shapes)
    mode: str = "technical"
    print_stats: dict[str, Any] | None = None
    # physical-properties readout (mass, com, inertia, tip, cost) for mode="physical"
    physical: dict[str, Any] | None = None


class SimulationPayload(BaseModel):
    """Motion-sim result: per-frame rigid poses keyed by leaf id, a run summary
    (worst clearance / collision frames), and the clearance ``require`` specs."""

    frames: list[dict[str, Any]] = Field(default_factory=list)
    summary: dict[str, Any] = Field(default_factory=dict)
    specs: list[dict[str, Any]] = Field(default_factory=list)


class ErrorPayload(BaseModel):
    message: str
    traceback: str = ""
    line: int | None = None


class StatusPayload(BaseModel):
    state: Literal["idle", "running", "ok", "error"]
    detail: str = ""
